code/dataobjects/Comment.php

```diff
@@ -267,59 +267,97 @@ public function canCreate($member = null) {
 	 * @return Boolean
 	 */
 	public function canView($member = null) {
-		if(!$member) $member = Member::currentUser();
+		$member = $this->getMember($member);
 
-		// Standard mechanism for accepting permission changes from decorators
 		$extended = $this->extendedCan('canView', $member);
-		if($extended !== null) return $extended;
+		if($extended !== null) {
+			return $extended;
+		}
+
+		if(Permission::checkMember($member, 'CMS_ACCESS_CommentAdmin')) {
+			return true;
+		}
 
-		// Allow admin
-		if(Permission::checkMember($member, 'CMS_ACCESS_CommentAdmin')) return true;
+		if($parent = $this->getParent()) {
+			return $parent->canView($member)
+				&& $parent->has_extension('CommentsExtension')
+				&& $parent->CommentsEnabled;
+		}
 
-		// Check if parent has comments and can be viewed
-		$parent = $this->getParent();
-		return $parent && $parent->ProvideComments && $parent->canView($member);
+		return false;
 	}
 
 	/**
-	 * Checks for "CMS_ACCESS_CommentAdmin" permission codes and
-	 * {@link canView()}.
+	 * Checks if the comment can be edited.
 	 *
-	 * @param Member $member
+	 * @param null|int|Member $member
 	 *
 	 * @return Boolean
 	 */
 	public function canEdit($member = null) {
-		if(!$member) $member = Member::currentUser();
+		$member = $this->getMember($member);
+
+		if(!$member) {
+			return false;
+		}
 
-		// Standard mechanism for accepting permission changes from decorators
 		$extended = $this->extendedCan('canEdit', $member);
-		if($extended !== null) return $extended;
+		if($extended !== null) {
+			return $extended;
+		}
 
-		if(!$this->canView($member)) return false;
+		if(Permission::checkMember($member, 'CMS_ACCESS_CommentAdmin')) {
```

Comment:
Moved up because all 'return true' permission checks should be promoted before 'return whatever expression' checks.